### ISSUE TITLE
Proxy build fails

### ISSUE THREAD

**Issue by preetiRK** [istio/proxy#695]
I am building the latest stable version of istio/proxy - **0.3.0** on **Ubuntu 16.04 x86_64 architecture**
Following dependencies have been installed 
   Bazel version : 0.8.0
   Go : go version go1.6.2 linux/amd64
   apt-get install pkg-config zip g++ zlib1g-dev unzip python wget openjdk-8-jdk libtool automake cmake git m4 golang-any

I see the following error when running the command `make build` :

> ERROR: /root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/mixerapi_git/mixer/v1/config/client/BUILD:4:1: Traceback (most recent call last):
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/mixerapi_git/mixer/v1/config/client/BUILD", line 4
                gogoslick_proto_library(name = "go_default_library", impor..."}, <7 more arguments>)
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/org_pubref_rules_protobuf/gogo/rules.bzl", line 123, in gogoslick_proto_library
                gogo_proto_library(name, langs, go_prefix, protos, <17 more arguments>)
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/org_pubref_rules_protobuf/gogo/rules.bzl", line 155, in gogo_proto_library
                go_proto_deps += PB_COMPILE_DEPS
trying to mutate a frozen object
ERROR: /root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/io_istio_api/mixer/v1/BUILD:4:1: Traceback (most recent call last):
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/io_istio_api/mixer/v1/BUILD", line 4
                gogoslick_proto_library(name = "go_default_library", impor..."}, <7 more arguments>)
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/org_pubref_rules_protobuf/gogo/rules.bzl", line 123, in gogoslick_proto_library
                gogo_proto_library(name, langs, go_prefix, protos, <17 more arguments>)
        File "/root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/org_pubref_rules_protobuf/gogo/rules.bzl", line 153, in gogo_proto_library
                go_proto_deps += GRPC_COMPILE_DEPS
trying to mutate a frozen object
ERROR: /root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/mixerapi_git/mixer/v1/config/client/BUILD:44:1: Target '@mixerapi_git//mixer/v1/config/client:api_spec.proto' contains an error and its package is in error and referenced by '@mixerapi_git//mixer/v1/config/client:cc_proto_genproto'
ERROR: /root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/mixerapi_git/mixer/v1/config/client/BUILD:44:1: Target '@mixerapi_git//mixer/v1/config/client:auth.proto' contains an error and its package is in error and referenced by '@mixerapi_git//mixer/v1/config/client:cc_proto_genproto'
ERROR: /root/.cache/bazel/_bazel_root/ef8bba518efe0190f2231cfcee9e87b1/external/mixerapi_git/mixer/v1/config/client/BUILD:44:1: Target '@mixerapi_git//mixer/v1/config/client:client_config.proto' contains an error and its package is in error and referenced by '@mixerapi_git//mixer/v1/config/client:cc_proto_genproto'

The full log can be seen at 
https://gist.github.com/preetiRK/570e67b2b62bc3a8d44ca2db5b6359eb

Any inputs will be appreciated.

**Comment by qiwzhang** [comment]
@ayj It seems:  istio/api:mixer/v1/config/client BUILD file has trouble with gogo stuff.    Could you help?

**Comment by ayj** [comment]
This seems to be already fixed in master. Looks like the mixerclient dependency SHA is bad in the 0.3.0 istio/proxy branch.

**Comment by rkpagadala** [comment]
@preetiRK We recently fixed master branch to build with 0.8.0 but that has not been back ported. Please use master branch with bazel 0.8.0 or 0.3.0 branch with bazel 0.7.0.

**Comment by preetiRK** [comment]
I tried building the 0.3.0 branch with bazel 0.7.0
It threw the following error

> ERROR: /root/.cache/bazel/_bazel_root/1647c3eda907e3d4d73b30afdf386635/external/envoy/bazel/repositories.bzl:232:13: no such package '@envoy_deps//': External dep build failed and referenced by '//external:googletest'.
ERROR: Analysis of target '//src/envoy/auth:jwt_test_lib' failed; build aborted: Loading failed.
INFO: Elapsed time: 663.112s
Makefile:25: recipe for target 'build' failed
make: *** [build] Error 1
The complete build log : https://gist.github.com/preetiRK/9121c61b9ff5f1bcb18e8ab0ca3ac068

I reran the `make build` command without doing any changes and without make clean. The build went ahead and next complained about permissions for file external/io_istio_istio/mixer/bin/generate_word_list.py 

> ERROR: /root/.cache/bazel/_bazel_root/1647c3eda907e3d4d73b30afdf386635/external/io_istio_istio/mixer/pkg/attribute/BUILD:32:1: Generating word list @io_istio_istio//mixer/pkg/attribute:global_list failed (Exit 1).
/bin/bash: external/io_istio_istio/mixer/bin/generate_word_list.py: Permission denied
INFO: Elapsed time: 2004.500s, Critical Path: 117.66s
Makefile:25: recipe for target 'build' failed
make: *** [build] Error 1

The complete build log : https://gist.github.com/preetiRK/9121c61b9ff5f1bcb18e8ab0ca3ac068#file-build-run-2

After changing the permissions to include execute rights, the build completes successfully
The complete build log :  https://gist.github.com/preetiRK/9121c61b9ff5f1bcb18e8ab0ca3ac068#file-build-run-3


The same behaviour was seen for the master branch of istio/proxy with bazel 0.8.0

**Comment by qiwzhang** [comment]
since your HOME is /root,  you may have to compile it as root user.

**Comment by kidiyoor** [comment]
I see the same  Permission denied error while building with 
- bazel 0.7 and release-0.3.0 branch (root user)
- bazel 0.8.1 and master branch (root user)

```
DEBUG: /root/.cache/bazel/_bazel_root/530ed184f85848d5a11a3a8203c74c1b/external/envoy/bazel/repositories.bzl:28:5: External dep build exited with return code: 0
INFO: Analysed 35 targets (224 packages loaded).
INFO: Found 35 targets...
ERROR: /root/.cache/bazel/_bazel_root/530ed184f85848d5a11a3a8203c74c1b/external/io_istio_istio/mixer/pkg/attribute/BUILD:32:1: Generating word list @io_istio_istio//mixer/pkg/attribute:global_list failed (Exit 1)
/bin/bash: external/io_istio_istio/mixer/bin/generate_word_list.py: Permission denied
```

**Comment by kidiyoor** [comment]
I use https://hub.docker.com/r/lyft/envoy-build-centos/ this docker image to build istio/proxy

**Comment by lizan** [comment]
Can you try to build with non-root user (even in a container)? That permission denied is due to https://github.com/bazelbuild/bazel/issues/3786.

**Comment by kidiyoor** [comment]
with non root user I see the following error

**master branch |  0.8.1 bazel**
```
/bin/sh ../../libtool  --tag=CXX   --mode=link /opt/rh/devtoolset-4/root/usr/bin/g++ -std=gnu++11 -Wno-deprecated-declarations -I /home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/src/c++11 -I /home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/src/c++11/lightstep -I/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_google_protobuf/src/ -g -O2 -L/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/bin/external/com_google_protobuf/ -lprotobuf -lprotobuf_lite  -o carrier_test carrier_test-carrier_test.o ../../src/c++11/liblightstep_core_cxx11.la
libtool: link: /opt/rh/devtoolset-4/root/usr/bin/g++ -std=gnu++11 -Wno-deprecated-declarations -I /home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/src/c++11 -I /home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/src/c++11/lightstep -I/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_google_protobuf/src/ -g -O2 -o carrier_test carrier_test-carrier_test.o  -L/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/bin/external/com_google_protobuf/ ../../src/c++11/.libs/liblightstep_core_cxx11.a -lprotobuf -lprotobuf_lite
make[3]: Entering directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test/c++11'
make[3]: Nothing to be done for `install-exec-am'.
make[3]: Nothing to be done for `install-data-am'.
make[3]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test/c++11'
make[2]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test/c++11'
make[2]: Entering directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test'
make[3]: Entering directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test'
make[3]: Nothing to be done for `install-exec-am'.
make[3]: Nothing to be done for `install-data-am'.
make[3]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test'
make[2]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test'
make[1]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build/test'
make[1]: Entering directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[2]: Entering directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[2]: Nothing to be done for `install-exec-am'.
make[2]: Nothing to be done for `install-data-am'.
make[2]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[1]: Leaving directory `/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/configure: line 5961: /usr/bin/file: No such file or directory
ERROR: /home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/external/envoy/source/common/router/BUILD:11:1: C++ compilation of rule '@envoy//source/common/router:config_lib' failed (Exit 4)
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
INFO: Elapsed time: 489.863s, Critical Path: 119.25s
FAILED: Build did NOT complete successfully
```


**release-0.3.0 branch | 0.7 bazel**
```
ERROR: missing input file '@envoy_deps//:thirdparty_build/lib/libgtest.a'.
ERROR: /home/bazeltest/proxy/src/envoy/auth/BUILD:89:1: //src/envoy/auth:jwt_test: missing input file '@envoy_deps//:thirdparty_build/lib/libgtest.a'.
ERROR: /home/bazeltest/proxy/src/envoy/auth/BUILD:89:1 1 input file(s) do not exist.
INFO: Elapsed time: 350.611s, Critical Path: 59.59s
```

**Comment by lizan** [comment]
`/home/bazeltest/.cache/bazel/_bazel_bazeltest/10430a57ac9debe95c2fe678d9bdb7b6/bazel-sandbox/7150962479503858874/execroot/__main__/external/com_github_lightstep_lightstep_tracer_cpp/configure: line 5961: /usr/bin/file: No such file or directory`

It is unusual that `/usr/bin/file` doesn't exist, does it exist in your container?

What is the base image do you use?

**Comment by preetiRK** [comment]
@qiwzhang  I did compile it as a root user.

@lizan I ran the commands as a non root. Here are my findings:

**0.3.0 branch with bazel 0.7.0** 
> ERROR: /home/test/.cache/bazel/_bazel_test/488ee1b4daca080a908b43d44da18f7b/external/envoy/bazel/repositories.bzl:232:13: no such package '@envoy_deps//': External dep build failed and referenced by '//external:tcmalloc_and_profiler'.
ERROR: Analysis of target '//src/envoy:envoy' failed; build aborted: Loading failed.
INFO: Elapsed time: 1179.778s
Makefile:25: recipe for target 'build' failed
make: *** [build] Error 1

Ran the make build command again and the logs were :

**Comment by preetiRK** [comment]
@qiwzhang I did compile it as a root user.

@lizan Today I tried the commands as a non root inside a ubuntu 16.04 container created using official ubuntu image. Here are the logs:

## 0.3.0 Branch with Bazel 0.7.0

> ERROR: /home/test/.cache/bazel/_bazel_test/488ee1b4daca080a908b43d44da18f7b/external/envoy/bazel/repositories.bzl:232:13: no such package '@envoy_deps//': External dep build failed and referenced by '//external:tcmalloc_and_profiler'.
ERROR: Analysis of target '//src/envoy:envoy' failed; build aborted: Loading failed.
INFO: Elapsed time: 1179.778s
Makefile:25: recipe for target 'build' failed
make: *** [build] Error 1

Re running the command 

> make[2]: Nothing to be done for 'install-exec-am'.
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/home/test/.cache/bazel/_bazel_test/488ee1b4daca080a908b43d44da18f7b/bazel-sandbox/2561378751436926545/execroot/__main__/bazel-out/local-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[1]: Leaving directory '/home/test/.cache/bazel/_bazel_test/488ee1b4daca080a908b43d44da18f7b/bazel-sandbox/2561378751436926545/execroot/__main__/bazel-out/local-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
ar: `u' modifier ignored since `D' is the default (see `U')
INFO: From MakeDeb tools/deb/istio-proxy_0.3-dev_amd64.deb:
/home/test/.cache/bazel/_bazel_test/488ee1b4daca080a908b43d44da18f7b/bazel-sandbox/5239778863042170363/execroot/__main__/bazel-out/host/bin/external/bazel_tools/tools/build_defs/pkg/make_deb.runfiles/bazel_tools/third_party/py/gflags/gflags/__init__.py:284: UserWarning: Flag architecture has a non-None default value; therefore, mark_flag_as_required will pass even if flag is not specified in the command line!
  'command line!' % flag_name)
INFO: Elapsed time: 2252.159s, Critical Path: 139.74 

## **Master branch with Bazel 0.8.0**

> ERROR: /home/test/.cache/bazel/_bazel_test/6167833dd35bb8e7499329dfd217bde9/external/envoy/bazel/repositories.bzl:232:13: no such package '@envoy_deps//': External dep build failed and referenced by '//external:googletest'
ERROR: Analysis of target '//src/envoy/auth:jwt_test' failed; build aborted: Loading failed
INFO: Elapsed time: 674.807s
FAILED: Build did NOT complete successfully (205 packages loaded)
Makefile:27: recipe for target 'build' failed
make: *** [build] Error 1

Rerunning make build command gives the following

> '/home/test/.cache/bazel/_bazel_test/6167833dd35bb8e7499329dfd217bde9/bazel-sandbox/4901097819202938709/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[2]: Nothing to be done for 'install-exec-am'.
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/home/test/.cache/bazel/_bazel_test/6167833dd35bb8e7499329dfd217bde9/bazel-sandbox/4901097819202938709/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
make[1]: Leaving directory '/home/test/.cache/bazel/_bazel_test/6167833dd35bb8e7499329dfd217bde9/bazel-sandbox/4901097819202938709/execroot/__main__/bazel-out/k8-fastbuild/genfiles/external/com_github_lightstep_lightstep_tracer_cpp/build'
ar: `u' modifier ignored since `D' is the default (see `U')
INFO: Elapsed time: 1696.457s, Critical Path: 85.23s
INFO: Build completed successfully, 1144 total actions

**Note:** the `INFO: Build completed successfully, ` statement was not seen for the 0.3.0 branch.

**Comment by qiwzhang** [comment]
Could you try install docker?  @JimmyCYJ run into similar problems.  The problem is gone after he installed docker.   @JimmyCYJ could you list the VM type and steps on your recent experience of building proxy?

**Comment by JimmyCYJ** [comment]
Here is the environment of the VM I use.
$ uname -a
Linux instance-2 4.10.0-42-generic #46~16.04.1-Ubuntu SMP Mon Dec 4 15:57:59 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
$ go version
go version go1.9.2 linux/amd64
$ docker version
Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:24:23 2017
 OS/Arch:      linux/amd64
$ bazel version
Build label: 0.8.1
Build target: bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Dec 5 14:29:45 2017 (1512484185)
Build timestamp: 1512484185
Build timestamp as int: 1512484185

I followed the instructions here https://github.com/istio/istio/blob/master/DEV-GUIDE.md. But I didn't install the Bazel buildifier tool and docker at first, which caused lots of errors when I ran bazel test src/envoy/mixer/... After I installed those tools it worked fine.

**Comment by preetiRK** [comment]
@JimmyCYJ  which branch of istio/proxy did you build? Did you use a non root user?

@qiwzhang I installed docker and buildifier tool. Also upgraded my go version to 1.9.2.
I tried building both the branches of proxy : master and 0.3.0.
The same error as earlier is seen.

**Comment by qiwzhang** [comment]
@preetiRK could you try to use GCE VM(Google Compute Engine) with base image Ubuntu 16.04 LTS to build proxy?

**Comment by JimmyCYJ** [comment]
@preetiRK I am not using root user, and I use a GCE VM to build. I fork istio/proxy and build, following the instructions here. https://github.com/istio/istio/blob/master/DEV-GUIDE.md#git-workflow.

**Comment by preetiRK** [comment]
@qiwzhang  I will try building on the GCE VM. 
Do you think that the issue I am facing are because of the image?

**Comment by qiwzhang** [comment]
My suspicious is some development packages, such as gcc, stdlib from the image.  I am pretty sure GCE VM works since many of us are using GCE vm to build proxy.   You can try to see the different between your image and GCE vm image.

**Comment by preetiRK** [comment]
Hi Folks.. Thanks for your response.

The timeout issue seen seems to be due to our docker container environment and is seen intermittently. 
I still see the issue sometimes. However on re-running the build command, the build completes and the test suite passes. So we have decided to ignore the timeout issue for now.
(note : I could build without installing docker and buildifier tool)

Closing the issue.
